moquette-0.16/broker/src/main/java/io/moquette/broker/security/DBAuthenticator.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2012-2018 The original author or authors
+ * ------------------------------------------------------
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Apache License v2.0 which accompanies this distribution.
+ *
+ * The Eclipse Public License is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * The Apache License v2.0 is available at
+ * http://www.opensource.org/licenses/apache2.0.php
+ *
+ * You may elect to redistribute this code under either of these licenses.
+ */
+
+package io.moquette.broker.security;
+
+import com.zaxxer.hikari.HikariDataSource;
+import io.moquette.BrokerConstants;
+import io.moquette.broker.config.IConfig;
+import org.apache.commons.codec.binary.Hex;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+/**
+ * Load user credentials from a SQL database. sql driver must be provided at runtime
+ */
+public class DBAuthenticator implements IAuthenticator {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DBAuthenticator.class);
+
+    private final MessageDigest messageDigest;
+    private HikariDataSource dataSource;
+    private String sqlQuery;
+
+    public DBAuthenticator(IConfig conf) {
+        this(conf.getProperty(BrokerConstants.DB_AUTHENTICATOR_DRIVER, ""),
+             conf.getProperty(BrokerConstants.DB_AUTHENTICATOR_URL, ""),
+             conf.getProperty(BrokerConstants.DB_AUTHENTICATOR_QUERY, ""),
+             conf.getProperty(BrokerConstants.DB_AUTHENTICATOR_DIGEST, ""));
+    }
+
+    /**
+     * provide authenticator from SQL database
+     *
+     * @param driver
+     *            : jdbc driver class like : "org.postgresql.Driver"
+     * @param jdbcUrl
+     *            : jdbc url like : "jdbc:postgresql://host:port/dbname"
+     * @param sqlQuery
+     *            : sql query like : "SELECT PASSWORD FROM USER WHERE LOGIN=?"
+     * @param digestMethod
+     *            : password encoding algorithm : "MD5", "SHA-1", "SHA-256"
+     */
+    public DBAuthenticator(String driver, String jdbcUrl, String sqlQuery, String digestMethod) {
+        this.sqlQuery = sqlQuery;
+        this.dataSource = new HikariDataSource();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code assigns a resource to a class field and there is no method that closes or returns this resource. This can result in a resource leak that slows down or crashes your system.

**Fix**
Create a method in this class that closes the following resource: *dataSource*. After you use this resource, use this method to prevent a resource leak.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).